main.go

```diff
@@ -177,91 +192,147 @@ func main() {
 			}
 		})
 	}
+
+	ctx := context.Background()
+
+	var cancel context.CancelFunc
+	if opts.Duration != 0 {
+		ctx, cancel = context.WithTimeout(ctx, opts.Duration)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+
 	{
-		t := time.NewTicker(opts.Period)
-		ctx, cancel := context.WithCancel(ctx)
+		g.Add(func() error {
+			return runWriter(ctx, opts, l, m)
+		}, func(_ error) {
+			cancel()
+		})
+	}
 
+	if opts.ReadEndpoint != nil {
 		g.Add(func() error {
-			level.Info(logger).Log("msg", "starting the remote-write client")
-			for {
-				select {
-				case <-t.C:
-					if err := write(ctx, endpointWrite, opts.Token, generate(opts.Labels)); err != nil {
-						level.Error(logger).Log("msg", "failed to make request", "err", err)
-					}
-				case <-ctx.Done():
-					return nil
-				}
-			}
+			return runReader(ctx, opts, l, m)
 		}, func(_ error) {
-			t.Stop()
 			cancel()
 		})
 	}
-	{
-		var cancel func()
 
-		if opts.Duration != 0 {
-			ctx, cancel = context.WithTimeout(ctx, opts.Duration)
+	if err := g.Run(); err != nil {
+		level.Error(l).Log("msg", "run group exited with error", "err", err)
+		os.Exit(1)
+	}
+
+	level.Info(l).Log("msg", "up completed its mission!")
+}
+
+func runReader(ctx context.Context, opts options, logger log.Logger, m metrics) error {
+	l := log.With(logger, "component", "reader")
+	level.Info(l).Log("msg", "starting the reader")
+
+	// Wait for at least one period before start reading metrics.
+	level.Info(l).Log("msg", "waiting for initial delay before querying for metrics")
+	select {
+	case <-ctx.Done():
+		return nil
+	case <-time.After(opts.InitialQueryDelay + opts.Period):
```

Comment:
this is not quite what we discussed. I think this introduces a regression to the place where the reader will wait for 2 writes to complete before the trying to read.

t=0 start writer timer
t=period write
t=period + delay start reader timer
t=2*period write
t=2*period + delay read